src/mds/Server.cc

```diff
@@ -6290,6 +6298,140 @@ void Server::handle_client_removexattr(MDRequestRef& mdr)
   journal_and_reply(mdr, cur, 0, le, new C_MDS_inode_update_finish(this, mdr, cur));
 }
 
+void Server::handle_client_getvxattr(MDRequestRef& mdr)
+{
+  const auto& req = mdr->client_request;
+  string name{req->get_path2()};
+
+  // is a ceph virtual xattr?
+  if (!is_ceph_vxattr(name)) {
+    respond_to_request(mdr, -CEPHFS_EINVAL);
+    return;
+  }
+
+  CInode *cur = rdlock_path_pin_ref(mdr, true, false);
+  if (!cur) {
+    return;
+  }
+
+  if (!cur->is_dir()) {
+    respond_to_request(mdr, -CEPHFS_ENOTDIR);
+    return;
+  }
+
+  CachedStackStringStream css;
+  int r = 0;
+  ceph::bufferlist bl;
+  // handle these vxattrs
+  if (name.substr(0, 15) == "ceph.dir.layout"sv) {
+    std::string layout_field;
+
+    struct layout_xattr_info_t {
+      const file_layout_t  layout;
+      const std::string    inheritance;
+
+      layout_xattr_info_t(const file_layout_t& l, const std::string& inh)
+        : layout(l), inheritance(inh) { }
+    };
+
+    auto is_default_layout = [](const file_layout_t& layout) -> bool {
+      auto default_layout = file_layout_t::get_default();
+      return (layout == default_layout);
+    };
+    auto get_inherited_layout = [&](CInode *cur) -> layout_xattr_info_t {
+      auto orig_in = cur;
+
+      while (cur) {
+        if (cur->get_projected_inode()->has_layout()) {
+	  auto& curr_layout = cur->get_projected_inode()->layout;
+	  if (is_default_layout(curr_layout)) {
+	    return {curr_layout, "default"s};
+	  }
+          if (cur == orig_in) {
+	      // we've found a new layout at this inode
+	      return {curr_layout, "new"s};
```

Comment:
Would "set" be more appropriate indicating that the layout is "set"?

Reply:
hmm ... it may be a personal preference; but when I see the word **new**, it makes me stop and take notice
the word **set** doesn't ring much of a bell
or maybe its just getting used to the terminology
since this is a user facing change, I'd want it to be impactful
I could just go ahead and change it to **set**; I'm not going to put up a fight :smiley:

Reply:
lets switch to using "set".

Reply:
ack; updated code to use an `enum` and replaced **"new"** with **"set"**
updated unit test to test for **"set"**